packages/react-router/modules/generatePath.js

```diff
@@ -0,0 +1,35 @@
+import pathToRegexp from 'path-to-regexp'
+
+const patternCache = {}
+const cacheLimit = 10000
+let cacheCount = 0
+
+const compileGenerator = (pattern) => {
+  const cacheKey = pattern
+  const cache = patternCache[cacheKey] || (patternCache[cacheKey] = {})
+
+  if (cache[pattern])
+    return cache[pattern]
+
+  const compiledGenerator = pathToRegexp.compile(pattern)
+
+  if (cacheCount < cacheLimit) {
+    cache[pattern] = compiledGenerator
+    cacheCount++
+  }
+
+  return compiledGenerator
+}
+
+/**
+ * Public API for generating a URL pathname from a pattern and parameters.
+ */
+const generatePath = (pattern = '/', params = {}) => {
+  if (pattern === '/') {
+    return pattern
```

Comment:
Obviously, this pattern does not contain any parameters. For an additional performance shortcut, we could assume that all strings that do not contain `*`, `(`, `)` or `:` are eligible to pass before creating a compile function.